### ISSUE TITLE
 Razor Class Library  in docker container referencing MVC throws  rzc generate exited with code 150

### ISSUE THREAD

**Issue by ChaosEngine** [dotnet/aspnetcore#16816]
### Describe the bug
Docker container (from image _mcr.microsoft.com/dotnet/core/sdk:3.0-buster_) with Razor Class Library project referencing nuget 

> <PackageReference Include="Microsoft.AspNetCore.Mvc" Version="2.2.0" />

 throws **error : rzc generate exited with code 150**

### To Reproduce
1. in command line execute

> docker run -it --rm mcr.microsoft.com/dotnet/core/sdk:3.0-buster

2. Inside container execute

> mkdir rcl && cd rcl && dotnet new razorclasslib

3. modify rcl.csproj in the following way to read

```
<Project Sdk="Microsoft.NET.Sdk.Razor">

  <PropertyGroup>
    <TargetFramework>netstandard2.0</TargetFramework>
    <RazorLangVersion>3.0</RazorLangVersion>
  </PropertyGroup>


  <ItemGroup>
    <PackageReference Include="Microsoft.AspNetCore.Components" Version="3.0.0" />
    <PackageReference Include="Microsoft.AspNetCore.Components.Web" Version="3.0.0" />
    <PackageReference Include="Microsoft.AspNetCore.Mvc" Version="2.2.0" />
  </ItemGroup>

</Project>
```

### Outcome
```
root@3a0f120b5bed:/rcl# dotnet build
Microsoft (R) Build Engine version 16.3.0+0f4c62fea for .NET Core
Copyright (C) Microsoft Corporation. All rights reserved.

  Restore completed in 16.53 ms for /rcl/rcl.csproj.
  It was not possible to find any compatible framework version
  The specified framework 'Microsoft.NETCore.App', version '2.0.9' was not found.
    - The following frameworks were found:
        3.0.0 at [/usr/share/dotnet/shared/Microsoft.NETCore.App]

  You can resolve the problem by installing the specified framework and/or SDK.

  The .NET Core frameworks can be found at:
    - https://aka.ms/dotnet-download
/usr/share/dotnet/sdk/3.0.100/Sdks/Microsoft.NET.Sdk.Razor/build/netstandard2.0/Microsoft.NET.Sdk.Razor.Component.targets(106,5): error : rzc generate exited with code 150. [/rcl/rcl.csproj]

Build FAILED.

/usr/share/dotnet/sdk/3.0.100/Sdks/Microsoft.NET.Sdk.Razor/build/netstandard2.0/Microsoft.NET.Sdk.Razor.Component.targets(106,5): error : rzc generate exited with code 150. [/rcl/rcl.csproj]
    0 Warning(s)
    1 Error(s)

Time Elapsed 00:00:20.54
```
### Expected outcome
Project builds

### Further technical details
- ASP.NET Core version

docker container run from image _mcr.microsoft.com/dotnet/core/sdk:3.0-buster_).
Same is happening with alpine based image: _mcr.microsoft.com/dotnet/core/sdk:3.0-alpine3.10_

### Include the output of `dotnet --info`

 dotnet --info
.NET Core SDK (reflecting any global.json):
 Version:   3.0.100
 Commit:    04339c3a26

Runtime Environment:
 OS Name:     debian
 OS Version:  10
 OS Platform: Linux
 RID:         debian.10-x64
 Base Path:   /usr/share/dotnet/sdk/3.0.100/

Host (useful for support):
  Version: 3.0.0
  Commit:  7d57652f33

.NET Core SDKs installed:
  3.0.100 [/usr/share/dotnet/sdk]

.NET Core runtimes installed:
  Microsoft.AspNetCore.App 3.0.0 [/usr/share/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 3.0.0 [/usr/share/dotnet/shared/Microsoft.NETCore.App]

To install additional .NET Core runtimes or SDKs:
  https://aka.ms/dotnet-download

**Comment by ChaosEngine** [comment]
Suprisingly setting env variable 

> DOTNET_ROLL_FORWARD_ON_NO_CANDIDATE_FX=2

 fixes problem, or lowering `<RazorLangVersion>2.0</RazorLangVersion>`

**Comment by mkArtakMSFT** [comment]
Thanks for contacting us, @ChaosEngine.
@NTaylorMullen can you please look into this? Thanks!

**Comment by NTaylorMullen** [comment]
Setting `<RazorLangVersion>3.0</RazorLangVersion>` to 3.0 for 2.0 applications is not supported sadly. Is there a reason why you're pushing that to 3.0?

**Comment by NTaylorMullen** [comment]
Sorry, I should have been more clear. Setting RazorLangVersion to 3.0 for applications targeting MVC 2.X, i.e. `<PackageReference Include="Microsoft.AspNetCore.Mvc" Version="2.2.0" />` is not supported. 

The reason why your roll forward ends up working I bet is that it's lifting your MVC version to be 3.0. If you remove that package reference entirely you'll still have access to MVC's 3.0 bits (we're part of the shared framework now) and won't have to set the rollforward piece.

**Comment by ChaosEngine** [comment]
You mean, to drop `<PackageReference Include="Microsoft.AspNetCore.Mvc" Version="2.2.0" /` and use `<FrameworkReference Include="Microsoft.AspNetCore.App" />` ?

Unfortunatelly this is not compiling even in VS2019 giving:

`:\Program Files\dotnet\sdk\3.0.100\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.Sdk.FrameworkReferenceResolution.targets(263,5): error NETSDK1073: The FrameworkReference 'Microsoft.AspNetCore.App' was not recognized`

```
<Project Sdk="Microsoft.NET.Sdk.Razor">

  <PropertyGroup>
    <TargetFramework>netstandard2.1</TargetFramework>
	  <GenerateEmbeddedFilesManifest>true</GenerateEmbeddedFilesManifest>
    <LangVersion>latest</LangVersion>
  </PropertyGroup>
  
  <ItemGroup>
    <FrameworkReference Include="Microsoft.AspNetCore.App" />
....
```

I've tried many combination and my Razor Class Lib requires MVC package to be present, (duh!)

**Comment by NTaylorMullen** [comment]
I'm suggesting don't have `<FrameworkReference Include="Microsoft.AspNetCore.App" />` either. The framework reference is implicit.

Assuming dropping Microsoft.AspNetCore.App doesn't work could you provide the output of `dotnet --info` and what specific VS version you're using?

**Comment by ChaosEngine** [comment]
Two VS2019 versions:
- VS2019 Community 16.3.7 (stable version)
- VS2019 Community 16.4.0 Preview 3 (preview version)

$ dotnet --info
.NET Core SDK (reflecting any global.json):
 Version:   3.1.100-preview2-014569
 Commit:    4bd5d24d87

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.18362
 OS Platform: Windows
 RID:         win10-x64
 Base Path:   C:\Program Files\dotnet\sdk\3.1.100-preview2-014569\

Host (useful for support):
  Version: 3.1.0-preview2.19525.6
  Commit:  5672978d91

.NET Core SDKs installed:
  3.0.100 [C:\Program Files\dotnet\sdk]
  3.1.100-preview2-014569 [C:\Program Files\dotnet\sdk]

.NET Core runtimes installed:
  Microsoft.AspNetCore.All 2.1.0 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.1.1 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.1.13 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.App 2.1.0 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.1.1 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.1.13 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.0.0 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.1.0-preview2.19528.8 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 2.1.13 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.0.0 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.1.0-preview2.19525.6 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.WindowsDesktop.App 3.0.0 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 3.1.0-preview2.19525.6 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]

To install additional .NET Core runtimes or SDKs:
  https://aka.ms/dotnet-download



Could the Preview VS2019 be the root cause of the issue?

**Comment by NTaylorMullen** [comment]
In the Visual Studio where you're seeing the issue could you open up Package Manager Console and type `dotnet --version`?

**Comment by ChaosEngine** [comment]
from VS 2019 stable:
```
Package Manager Console Host Version 5.3.1.6268

Type 'get-help NuGet' to see all available NuGet commands.

PM> dotnet --version
3.1.100-preview2-014569
PM> 
```

**Comment by NTaylorMullen** [comment]
Ahh so you have the enable preview sdk flag on I presume? 3.1 is not supported in ASP.NET projects in 16.3. Do you have the same issue in your 16.4 instance?

**Comment by ChaosEngine** [comment]
Yes, same on 16.4 version and built from CLI. Here's 16.3
![image](https://user-images.githubusercontent.com/1271428/68241854-bc301a80-000f-11ea-8eeb-d333f3853d54.png)

...and 16.4 preview options:
![image](https://user-images.githubusercontent.com/1271428/68242039-1204c280-0010-11ea-8c52-aa4a50ccf8f7.png)

Well, months ago I was enabling some preview flags AFAIR ;-)

**Comment by NTaylorMullen** [comment]
Hmm, i'm truly confused. At this point I imagine your SDK or VS is in a broken state. Do you see the issue at command line via `dotnet build`?

**Comment by ChaosEngine** [comment]
yes, but currently I'm more interested in proper build the project in docker container witn 3,0 sdk. But nevertheless her'es the cli win 10 repro:

```
E:\Projects\RazorClassLibrary1>type RazorClassLibrary1.csproj
´╗┐<Project Sdk="Microsoft.NET.Sdk.Razor">

  <PropertyGroup>
    <TargetFramework>netstandard2.1</TargetFramework>
    <!--<RazorLangVersion>3.0</RazorLangVersion>-->
  </PropertyGroup>

  <ItemGroup>
    <FrameworkReference Include="Microsoft.AspNetCore.App" />
  </ItemGroup>

  <ItemGroup>
    <PackageReference Include="Microsoft.AspNetCore.Components" Version="3.0.0" />
    <PackageReference Include="Microsoft.AspNetCore.Components.Web" Version="3.0.0" />
    <PackageReference Include="Microsoft.AspNetCore.Mvc" Version="2.2.0" />
  </ItemGroup>

</Project>

E:\Projects\RazorClassLibrary1>dotnet --list-sdks
3.0.100 [C:\Program Files\dotnet\sdk]
3.1.100-preview2-014569 [C:\Program Files\dotnet\sdk]

E:\Projects\RazorClassLibrary1>dotnet build
Microsoft (R) Build Engine 16.4.0-preview-19528-01+3745451f4 dla platformy .NET Core
Copyright (C) Microsoft Corporation. Wszelkie prawa zastrzeżone.

  Ukończono przywracanie w elemencie 35,73 ms (E:\Projects\RazorClassLibrary1\RazorClassLibrary1.csproj).
  Korzystasz z wersji zapoznawczej programu .NET Core. Zobacz: https://aka.ms/dotnet-core-preview
C:\Program Files\dotnet\sdk\3.1.100-preview2-014569\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.Sdk.FrameworkReferenceResolution.targets(283,5): error NETSDK1073: Nie rozpoznano elementu FrameworkReference "Microsoft.AspNetCore.App" [E:\Projects\RazorClassLibrary1\RazorClassLibrary1.csproj]

Kompilacja NIE POWIODŁA SIĘ.

C:\Program Files\dotnet\sdk\3.1.100-preview2-014569\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.Sdk.FrameworkReferenceResolution.targets(283,5): error NETSDK1073: Nie rozpoznano elementu FrameworkReference "Microsoft.AspNetCore.App" [E:\Projects\RazorClassLibrary1\RazorClassLibrary1.csproj]
    Ostrzeżenia: 0
    Liczba błędów: 1

Czas, który upłynął: 00:00:00.58
```

BTW. I'm using polish pl-PL locale

**Comment by NTaylorMullen** [comment]
If you do
```
<Project Sdk="Microsoft.NET.Sdk.Razor">

  <PropertyGroup>
    <TargetFramework>netstandard2.1</TargetFramework>
    <RazorLangVersion>3.0</RazorLangVersion>
  </PropertyGroup>


  <ItemGroup>
    <PackageReference Include="Microsoft.AspNetCore.Components" Version="3.0.0" />
    <PackageReference Include="Microsoft.AspNetCore.Components.Web" Version="3.0.0" />
  </ItemGroup>

</Project>
```

at the command line does it build?

**Comment by ChaosEngine** [comment]
Clean repro: in docker container mcr.microsoft.com/dotnet/core/sdk:3.0-buster with dotnet new razorclasslib: **builds**
in stable VS 2019 16.3.7 htowu new project rcl: **builds**

**Comment by vishalvips** [comment]
/opt/dotnetsdk/sdk/NuGetFallbackFolder/microsoft.aspnetcore.razor.design/2.2.0/build/netstandard2.0/Microsoft.AspNetCore.Razor.Design.CodeGeneration.targets(79,5): error : rzc discover exited with code 134. [/opt/dotnetsdk/sdk/2.2.207/myapp/myapp.csproj]

The build failed. Please fix the build errors and run again.


My Version info :

.NET Core SDK (reflecting any global.json):
 Version:   2.2.207
 Commit:    f7b16a2cd5

Runtime Environment:
 OS Name:     rhel
 OS Version:  6
 OS Platform: Linux
 RID:         rhel.6-x64
 Base Path:   /opt/dotnetsdk/sdk/2.2.207/

Host (useful for support):
  Version: 2.2.8
  Commit:  b9aa1abc51

.NET Core SDKs installed:
  2.2.207 [/opt/dotnetsdk/sdk]

.NET Core runtimes installed:
  Microsoft.AspNetCore.All 2.2.8 [/opt/dotnetsdk/shared/Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.App 2.2.8 [/opt/dotnetsdk/shared/Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 2.2.8 [/opt/dotnetsdk/shared/Microsoft.NETCore.App]

To install additional .NET Core runtimes or SDKs:
  https://aka.ms/dotnet-download

**Comment by mkArtakMSFT** [comment]
Thanks for contacting us. This is already fixed in 3.1.
